(SLV-730) Create PuppetDB queue depth analyzer

johnduarte opened the pull request.

johnduarte committed:
(SLV-730) Add PmcMiller::Reader class

Add PmcMiller::Reader class.  Add the ability to set a service name.

johnduarte committed:
(SLV-730) Implement reading of JSON for reader

This commit implements the ability for the PmcMiller::Reader class
to read JSON files.

Spec test is added for this functionality along with supporting fixture
tree.

johnduarte committed:
(SLV-730) Add travis-ci status shield to README

johnduarte committed:
(SLV-730) Add pry-byebug to gemspec

johnduarte committed:
(SLV-730) Create Result/Results objects

This commit creates PmcMiller::Result and PmcMiller::Results objects to
manage the result/results for a given query.

johnduarte committed:
(SLV-730) Add PuppetDB::QueueDepth object

This commit adds the PmcMiller::PuppetDB::QueueDepth class to implement
the interface for queue depth analysis of puppetdb
puppet_metrics_collector data.

johnduarte committed:
(SLV-730) Use Time rather than DateTime

This commit replaces DateTime object instances with Time instances.
DateTime calculations return fractions of a day rather than seconds
which is what we need to calculate the slope of the queue_depth
increases.

johnduarte committed:
(SLV-730) Sort json files

johnduarte commented:
https://tickets.puppetlabs.com/browse/SLV-745 opened to track adding the ability to pass multiple lookup values to reader.
https://tickets.puppetlabs.com/browse/SLV-746 opened to track splitting the calculations and analyses into discrete classes.

johnduarte committed:
(SLV-730) Rename Result(s) to DataPoint/Data

This commit renames the Result and Results classes to DataPoint and Data
respectively.  After review the previous names were determined to not be
related to the desired "result" which as a method on an analysis object
should return the numeric results of the analysis.  Since these items are
simply the data gathered from the puppet-metric-collector without
modification, they have been renamed to make that relationship clearer.

johnduarte committed:
(SLV-730) Change QueueDepth to take data

This commit changes the interface for the QueueDepth class to take a
Data object rather than a Reader.  The user now has to read the data on
their own, but the class now only analyzes the data rather than wrapping
the reading and analysis.

johnduarte committed:
(SLV-730) Update README

johnduarte committed:
(SLV-730) Add yard docs for PmcMiller::Reader

johnduarte committed:
(SLV-730) Add yard docs for PmcMiller::Data

johnduarte committed:
(SLV-730) Yard docs PmcMiller::PuppetDB::QueueDepth

johnduarte committed:
(SLV-730) Replace key_map with nested_hash_value

To prevent a massive explosion of hard-coded hash maps to get to the
deeply nested keys in the puppet-metric-collector data, this commit
replaces the `key_map` method with a `nested_hash_value` method.  Now
the `PmcMiller::Reader.read` method will recursively search the
appropriate JSON files for a key that matches the given string.  It
returns the value associated with the first found key in the hash
derived from the JSON data.

RandellP commented on README.md:
```diff
@@ -1,15 +1,17 @@
-Welcome to your new gem! In this directory, you'll find the files you need to be able to package up your Ruby library into a gem. Put your Ruby code in the file `lib/pmc_miller`. To experiment with that code, run `bin/console` for an interactive prompt.
 
-TODO: Delete this and the text above, and describe your gem
+PmcMiller provides an interface to read and analyze data gathered by
```
and create new derived data... as well as add data... like we want to read info from gatling (like KO's or an response times) and then use this to add that data to subdir of the PMC dir so we can then import all of it into grafana in one step.

johnduarte replied:
That is far beyond the scope of SLV-730.  That functionality sounds like another gem entirely.  PmcBaker perhaps. :)

RandellP replied:
All you would need is a writer. Should be one class similar to the reader.  And it would make more sense to add it to this gem so you can reuse existing code that would be shared.  It is certainly outside the scope of this ticket though...

RandellP commented on README.md:
```diff
@@ -20,16 +22,89 @@ Or install it yourself as:
+
+reader = PmcMiller::Reader.new("/path/to/puppet-metrics-collector")
+reader.service = :puppetdb
+data = reader.read("queue_depth")
```
Any reason we don't just pass in the service with the key to be read?  Seems like an arbitrary extra line from the user perspective.

johnduarte replied:
This was just my mental model of the object.  I thought it might provide preventing from having to pass the service for other methods that would require its context which would always be needed (e.g. reader.foo(service, bar)). If the `read` method is the only method, then it makes sense to add a service parameter.

The other option is to add the service to the `initialize` method so that it is configured at the time of creation.
```
db_reader = PmcMiller::Reader.new("/path/to/puppet-metrics-collector", :puppetdb)
srv_reader = PmcMiller::Reader.new("/path/to/puppet-metrics-collector", :puppetserver)
qd_data = db_reader.read("queue_depth")
free_jruby_data = srv_reader.read("average-free-jrubies")
```

RandellP replied:
I don't think you want to add it to initialize, because then you have to create a reader per service... instead of one reader to read from any service.  And there isn't any reason to have more than one reader.  
And just in case my old suggestion of munging it with the key is in your head... don't... it was a bad idea... :)  having it as a param to the read seems like the best approach.

johnduarte replied:
The reader class has been refactored not to have a service attribute.  Thus the reader object is no longer tied to a particular service at any time.  The `read` method now takes a service and key to attempt to read.

RandellP commented on README.md:
````diff
@@ -20,16 +22,89 @@ Or install it yourself as:
+```
+
+The data object returned is an array of PmcMiller::DataPoint objects.  These
+objects are a struct consisting of the `Time` and value of the data point for
````
Why an array of structs versus a hash?   I am guess more efficient in some way. 
But it has some functional differences as well.  An array of structs would allow for multiple data points from the same timestamp... where as the hash wouldn't.  (this is good in that it preserves the raw data, and bad in that anything using it will have to deal with that possibility).  Also an array will have an explicit order, which likely is the same as the files were processed, which is probably earliest to latest.  That could be good if it can be counted on... and if so, should say so here so the user knows it is essentially pre-sorted.

johnduarte replied:
As you stated, I wanted to preserve the data points gathered from the JSON, much like the [current documentation describes](https://github.com/puppetlabs/puppetlabs-puppet_metrics_collector#grepping-metrics). The array allows the data object to be sorted in any way the user is interested in getting the information.

RandellP commented on README.md:
```diff
@@ -20,16 +22,89 @@ Or install it yourself as:
+: JSON string of the previous three methods.
+
+
+#### Example Usage
```
The reader didn't have this subheading... I like it personally, but either way, should be consistent across the classes.

johnduarte replied:
Heading removed.

RandellP commented on lib/pmc_miller/puppetdb/queue_depth.rb:
```diff
@@ -0,0 +1,124 @@
+      def slope(dpoint1, dpoint2)
+        unless @data.respond_to?(dpoint1) &&
+               @data.respond_to?(dpoint2) &&
+               %w[first mid last].include?(dpoint1) &&
```
Are these last two checks redundant?  Seems like if the object responds to the method, checking if it is in a hardcoded list adds a maintenance burden to keep the list up to date...
I suppose a person could input "slope" and that would pass the first test but not the second.  But if that is a concern it might be better to make a method for checking those inputs... it could just call the method and ensure the return is of type data struct.  
Alternatively just hardcoding a list in an attribute or something, and referring to it allow should be enough.  
If a proc like this was called on many data points the efficiency of the checking would matter.  (and of course this implementation is meant to be a model for others...)

johnduarte replied:
Call me paranoid, but I don't want the data object to be able take arbitrary Array methods in this context.  Without the final check for the expected methods, the `slope` method would try to digest the following methods with unexpected results:
```
irb(main):001:0> d = PmcMiller::Data.new
irb(main):002:0> d.respond_to?(:<=>)
=> true
irb(main):003:0> d.respond_to?(:<<)
=> true
```

RandellP replied:
yeah, so then just do the hardcoded list check... though put the list in a variable or something.

RandellP commented on lib/pmc_miller/puppetdb/queue_depth.rb:
```diff
@@ -0,0 +1,124 @@
+          return
+        end
+
+        delta_x = @data.public_send(dpoint2).value - @data.public_send(dpoint1).value
```
I think I know why we had trouble following each other on this piece.  I was thinking of slope like a best fit line, which would consider all points between.  And you are using it for just the slope between any two points... which would ignore any spikes in the middle.  This of course would be super fragile.  If the que depth just happened to be zero are the mid point and the first(which is should be in theory), and non-zero at the end...  that would be a fail.  And in theory it should always be zero at the end if the data goes out far enough, which would make it always pass...  Either way fragile.  But I think you said we would work on the algorithm separately.

RandellP commented on lib/pmc_miller/puppetdb/queue_depth.rb:
```diff
@@ -0,0 +1,124 @@
+      #
+      # The average rate of change (slope) is calculated for the given points
+      # in the data set.
+      #
```
It's obvious from the code details, but should be mentioned in the comment, that the time order of the dpoints matters... must be from earliest to latest...

johnduarte replied:
Yes.  I think you have also spotted a possible bug.  The slope method should ensure that the data is sorted properly and should not assume that something else has done so in advance.

johnduarte replied:
Comment and method updated to ensure the data is sorted by time ascending prior to calculation.

RandellP commented on lib/pmc_miller/puppetdb/queue_depth.rb:
```diff
@@ -0,0 +1,124 @@
+      #
+      # @return [Float]
+      #
+      def rate_of_change
```
Value should be cached in a attribute or something to prevent doing the whole calculation multiple times (which it currently does).

johnduarte replied:
Memoization added for this method.

RandellP commented on lib/pmc_miller/reader.rb:
```diff
@@ -0,0 +1,94 @@
+    ##
+    # Creates a new reader that will use the given path as its data source.
+    #
+    # @param path [String] File system path to puppet-metrics-collector data
```
should probably specifiy that it is the root dir of the PMC data... since all the subdirs meet the current criteria documented.

johnduarte replied:
Updated.

RandellP commented on lib/pmc_miller/reader.rb:
```diff
@@ -0,0 +1,94 @@
+    #
+    # @return [PmcMiller::Data] Data collection found
+    #
+    def read(key)
```
I think that either the "key" should include the service in some way (like service::key or something), or this method should accept the service as a parameter.

RandellP replied:
I think read should just take a service... it's required to do the job, but only needed to do read... so I can't think of a reason to separate it out.  Though maybe we discussed this and I am already forgetting?

RandellP commented on lib/pmc_miller/reader.rb:
```diff
@@ -0,0 +1,94 @@
+    # @return [PmcMiller::Data] Data collection found
+    #
+    def read(key)
+      return unless %i[puppetserver puppetdb].include? @service
```
This should be a directory existence check instead of a hardcoded list check.

johnduarte replied:
Edge case: What would you expect the behavior to be for the following cases:
* The reader's service is assigned to a valid service that pmc failed to collect data for?  There would be no directory.  I would expect an empty dataset to be returned.
* The reader's service is assigned to a service that pmc does not support?  There would be no directory.  I would expect an exception to be thrown.

These cases lead me to create a whitelist of services rather than relying on the directory structure.  I am interested in your thoughts @RandellP .

RandellP replied:
If PMC failed to collect data, there should still be a dir, it is created by the module in a way that is not connected to data collection I believe.  But I think it should raise rather than an empty dataset.  Cause there are multiple reason why it has no data, and the calling method can't figure out which it is from an empty dataset.  But a raised error can be caught, processed, and decisions made.  Like the key may not be in the data, there may be no data, there may be no dir...  heck, the jsons maybe invalid...  Lots of things could go wrong.  In most cases, the calling method will want to catch it and continue on in some way, but store up a list of issues to report out at the end.   All that said, I am not married to any part of what I said... :)

RandellP commented on lib/pmc_miller/reader.rb:
```diff
@@ -0,0 +1,94 @@
+    # @param obj [Object] Object to recursively search
+    # @param key [Symbol, string] Key to search for
+    #
+    # @return First found value associated with key, nil if not found
```
I don't love the "first found" aspect here.  Seems error prone.  Especially when many of the keys will be the same like "mean"...  And when of course the order of the keys in the json aren't guaranteed to remain the same.  But that could be a separate ticket.  probably a config file mapping short names to full hierarchy of the key.

johnduarte replied:
This has been reverted and to re-establish the key_map logic for only honoring the `:queue_depth` query.

RandellP commented on spec/pmc_miller_puppetdb_queue_depth_spec.rb:
```diff
@@ -0,0 +1,60 @@
+require "pmc_miller/reader"
+require "pmc_miller/puppetdb/queue_depth"
+
+require "pry"
```
Did you mean to leave this in?

johnduarte replied:
Deleted.

RandellP commented on spec/pmc_miller_puppetdb_queue_depth_spec.rb:
```diff
@@ -0,0 +1,60 @@
+    subject = PmcMiller::PuppetDB::QueueDepth.new(data)
+
+    it "gathers data of puppetdb queue depth" do
+      # expected valuesextracted from fixture json file
```
s/valuesextracted/values extracted/

johnduarte replied:
updated.

RandellP commented on spec/pmc_miller_puppetdb_queue_depth_spec.rb:
```diff
@@ -0,0 +1,60 @@
+    it "fails" do
+      expect(subject.summary).to eq("fail")
+    end
+  end
```
should we add a rate of chance check to each of these.  In theory we should be checking all the outputs really if we are doing integration style spec tests no?

johnduarte replied:
Spec test additions added.

RandellP commented on spec/pmc_miller_reader_spec.rb:
```diff
@@ -0,0 +1,18 @@
+    reader = PmcMiller::Reader.new(fixtures_dir)
+    reader.service = :puppetdb
+    data = reader.read("queue_depth")
+    expect(data).to be_an_instance_of(PmcMiller::Data)
```
From a user perspective, I think we care more that it actually contain the data from the files than that the type is correct...

johnduarte replied:
...which I left as an exercise for the objects using the data.  This is my TDD test for bootstrapping the reader, for which I am only concerned that he correct objects were generated.

RandellP replied:
hm, you mean like, the spec test for quedepth would confirm that the reader actually gave back the right summary and results?  (which I just remembered it does).  That seem backwards actually.  Cause there are more then one analyzers, so doing it that way means they all check that the right data got read?  I would think the reader confirms that data got read, and the analyzers assume it did, and only check that the calculation is correct.

johnduarte committed:
Revert "(SLV-730) Replace key_map with nested_hash_value"

This reverts commit e3afb2bec430b4eba041fe5dc250d239743909b9.

johnduarte committed:
(SLV-730) Misc fixups

johnduarte committed:
(SLV-730) Sort queue_depth data for slope

This commit ensures that the data used by the
`PmcMiller::PuppetDB::QueueDepth.slope` is sorted by time in ascending
order prior to calculating the slope value.

billclaytor commented on README.md:
```diff
@@ -20,16 +22,88 @@ Or install it yourself as:
+
+The data object returned is an array of PmcMiller::DataPoint objects.  These
+objects are a struct consisting of the `Time` and value of the data point for
+the key that was use for the lookup.  From our previous example, the Data
```
```suggestion
the key that was used for the lookup.  From our previous example, the Data
```

johnduarte committed:
Update README.md

Co-Authored-By: Bill Claytor <35778171+billclaytor@users.noreply.github.com>

billclaytor commented on README.md:
```diff
@@ -20,16 +22,88 @@ Or install it yourself as:
 ## Development
 
-After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.
+After checking out the repo, run `bin/setup` to install dependencies. Then, run
```
I noticed that the `bin/setup` script is not included in the PR, but:
1) It would be helpful if it included comments
2) I always prefer to do `bundle install --path vendor/bundle`. Is this possible / preferable?

johnduarte replied:
The `--path` flag will be removed from bundler in v3 and as of v2.1 emits deprecation warnings indicating that it should not be used. See the [bundler docs](https://github.com/rubygems/bundler/blob/d4993be66fa2e76b3ca00ea56a51ecab5478b726/UPGRADING.md#bundler-3) for more info.

johnduarte committed:
(SLV-730) Convert Data to Array object

The Data object created too much overhead just to support a "mid"
method.  This commit converts the Data object to a built in Array.  The
mid functionality is now baked into the slope calls win QueueDepth.

billclaytor commented on lib/pmc_miller/reader.rb:
```diff
@@ -0,0 +1,95 @@
+    private
+
+    ##
+    # Fully defined data key structure assocated with given key.
```
```suggestion
    # Fully defined data key structure associated with given key.
```

johnduarte committed:
Update lib/pmc_miller/reader.rb

Co-Authored-By: Bill Claytor <35778171+billclaytor@users.noreply.github.com>

johnduarte committed:
(SLV-730) QueueDepth Add rate_of_change spec

This commit adds rate_of_change validation to the spec tests for
PmcMiller::PuppetDB::QueueDepth.

johnduarte committed:
(SLV-730) Memoize rate_of_change

This commit memoizes the rate_of_change method for
PmcMiller::PuppetDB::QueueDepth to prevent redundant calls to the
calculation.

RandellP commented on lib/pmc_miller/puppetdb/queue_depth.rb:
```diff
@@ -0,0 +1,122 @@
+      def rate_of_change
+        return @r_o_c if defined?(@r_o_c)
+
+        @r_o_c = (first_half_slope + last_half_slope) / 2.0
```
This class attribute should be documented with the class up above.  And I just am not a fan of names like r_o_c.  I think we can afford the extra characters... :)

johnduarte replied:
I have renamed the instance variable so that it is less cryptic.  However, it is not an attribute and is not accessible as such.  It is intended as an internal instance variable to cache the rate_of_change calculation.  For that reason, I would argue that it should not be publicly documented as an attribute of the class.

johnduarte committed:
(SLV-730) Introspect services from PMC path

This commit changes the available services to query from a hard-coded
list of symbols to introspecting them from the directory structure of
the PMC path.

johnduarte committed:
(SLV-730) Remove service as object attribute

This commit removes service as an object attribute and implements it as
a parameter given to the `read` method.

johnduarte committed:
(SLV-730) Raise when no service:key map

This commit updates the `read` method to raise an exception if the given
service:key combination does not have a data map into the PMC data.

johnduarte committed:
(SLV-730) Change read key to string

This commit changes the `key` parameter for the `read` method to a
String rather than a symbol to make it consistent with the `service`
parameter.

johnduarte committed:
(SLV-730) Rename r_o_c instance variable

johnduarte commented:
@RandellP I have attempted to address your concerns.  Please let me know if I have not fully satisfied them.  Thanks.

RandellP commented on lib/pmc_miller/puppetdb/queue_depth.rb:
```diff
@@ -0,0 +1,122 @@
+      # @return [Float]
+      #
+      def slope(dpoint1, dpoint2)
+        @data = @data.sort_by(&:time)
```
hm, this means the data will get sorted at least twice... (slope is currently called twice).  That seems bad, though maybe if it is already sorted the cost to sort the same way is really low?

johnduarte replied:
This has been moved into the initialize method of the analyzer object.

johnduarte committed:
(SLV-730) Sort data when QueueDepth is created

This commit moves the sorting of the data from the slope calculation to
the initializer of the `PmcMiller::PuppetDB::QueueDepth` object.  This
ensures that any methods using the data will have sorted data and that
the sorting operation is only executed once.

billclaytor reviewed (approved):
Awesome!
